### PR TITLE
deny s3:DeleteBucket for cdo buckets

### PR DESCRIPTION
**Pull request by cat5inthecradle**
Deny the deletion of critical S3 buckets.

This would need to be manually added to several roles (GoogleUser, for example) as they are not managed via code.

Alternatively - we could add this Deny to the DevPermissions policy, which is used as a permissions boundary elsewhere. We use that boundary for cloudformation roles, so it's possible that denying deletebucket could cause problems there.

## Deployment strategy

manually deploy via README

## Follow-up work

Apply role to manually created Roles

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: aws/cloudformation/iam.yml.erb [hunk 1]
```diff
@@ -147,6 +147,19 @@ Resources:
...
+    Type: AWS::IAM::ManagedPolicy
+    Properties:
+      ManagedPolicyName: DenyS3Delete
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Deny
+            Action:
+              - s3:DeleteBucket
+            Resource: "arn:aws:s3:::cdo-*"
```
Comment (cat5inthecradle) [inline aws/cloudformation/iam.yml.erb:161]: Are there buckets outside this prefix we should include?

Should we just use '*' and deny for all?

**Comment by cat5inthecradle** [comment]
I think we should find a higher-level place to do this, if possible, rather than attempting to remember to add it for all roles.

File: aws/cloudformation/iam.yml.erb (hunk 1 above)
Comment (sureshc) [reply aws/cloudformation/iam.yml.erb:161]: Is there a way we can refine this to just production buckets (maybe using the `environment` tag)? I’m thinking about Javabuilder, where each Stack has its own buckets to store user generated content and an engineer and their CodePipeline needs to be able to delete the buckets when they de-provision their development Stack. While we currently use buckets that are shared across all environments to store student project code & assets, we hopefully will implement separate buckets per Stack in the future.

Comment (cat5inthecradle) [reply aws/cloudformation/iam.yml.erb:161]: Related PR on that topic - tagging might be a good solution here.
https://github.com/code-dot-org/javabuilder/pull/394

**Pull request closed without merging**